compiler/sem/semtypinst.nim

```diff
@@ -423,72 +414,73 @@ proc instCopyType*(cl: var TReplTypeVars, t: PType): PType =
     result.flags.excl tfHasAsgn
 
 proc handleGenericInvocation(cl: var TReplTypeVars, t: PType): PType =
-  # tyGenericInvocation[A, tyGenericInvocation[A, B]]
-  # is difficult to handle:
-  var body = t[0]
+  ## Evaluates a type application (a ``tyGenericInvocation``) and returns
+  ## the resulting ``tyGenericInst``. The result is cached, and evaluating
+  ## the type application again won't create a new instance, but will instead
+  ## return the cached instance.
+  let body = t.base
   cl.c.config.internalAssert(body.kind == tyGenericBody, cl.info, "no generic body")
   var header = t
-  # search for some instantiation here:
-  result = searchInstTypes(cl.c.graph, t)
 
-  if result != nil and sameFlags(result, t):
-    when defined(reportCacheHits):
-      echo "Generic instantiation cached ", typeToString(result), " for ", typeToString(t)
-    return
+  let oldSkipTypedesc = cl.skipTypedesc
+  cl.skipTypedesc = true
+
+  # instantiate all invocation arguments:
   for i in 1..<t.len:
-    var x = t[i]
-    if x.kind in {tyGenericParam}:
-      x = lookupTypeVar(cl, x)
-      if x != nil:
-        if header == t: header = instCopyType(cl, t)
-        header[i] = x
-        propagateToOwner(header, x)
-    else:
+    let it = t[i]
+    var x =
+      if it.kind == tyGenericParam:
+        lookupTypeVar(cl, it)
+      else:
+        replaceTypeVarsT(cl, it)
+
+    if x.kind == tyTypeDesc:
+      # we want the unwrapped type in argument positions
+      x = x.base
+
+    if x != it:
+      if header == t:
+        # optimization: only create a copy when something changes. We only
+        # use the header for the cache lookup, so an exact replica suffices
+        header = exactReplica(t)
+      header[i] = x
       propagateToOwner(header, x)
 
-  if header != t:
-    # search again after first pass:
-    result = searchInstTypes(cl.c.graph, header)
-    if result != nil and sameFlags(result, t):
-      when defined(reportCacheHits):
-        echo "Generic instantiation cached ", typeToString(result), " for ",
-          typeToString(t), " header ", typeToString(header)
-      return
-  else:
-    header = instCopyType(cl, t)
+  cl.skipTypedesc = oldSkipTypedesc
 
+  # search the instance cache for the type:
+  result = searchInstTypes(cl.c.graph, header)
+  if result != nil:
+    when defined(reportCacheHits):
+      echo "Generic instantiation cached ", typeToString(result), " for ",
+        typeToString(t), " header ", typeToString(header)
+    return
+
+  # not cached yet
   result = newType(tyGenericInst, nextTypeId(cl.c.idgen), t[0].owner)
-  result.flags = header.flags
+  # make sure the 'not-nil' flag (which is essentialy a type modifier) is
+  # already present on the instance prior to caching
+  result.flags = header.flags * {tfNotNil}
```

Comment:
sorry, if this is clarified elsehwere but are we assuming that the instantiated type is not nil because we are instantiating it therefore any function pointer to said type would resolve?

Reply:
The `tfNotNil` flag is the only flag contributing to type equality that can currently be set on the invocation, and these flags have to be present on the instance prior to recursing, otherwise there would be issues with recursive generics.

An example of where problems would occur (`a` would get the same instance as `y`, which would be wrong):  
```nim
type Generic[T] = ref object
  x: Generic[T]

var y: Generic[int] not nil
```

Reply:
In order for the code to be a bit more future-proof, I'm going to change `{tfNotNil}` to `eqTypeFlags`.